dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/converter/TrackerBundleParamsConverter.java

```diff
@@ -84,174 +82,238 @@
     extends
     StdConverter<TrackerBundleParams, TrackerBundleParams>
 {
+
+    /**
+     * Iterates over the collections of a dataBundle. If any objects in those collections have objects nested within
+     * them, they are extracted. For each object we process, we make sure all references are valid as well.
+     *
+     * @param dataBundle containing collections to check and update.
+     * @return a dataBundle with a flattened data structure, and valid uid references.
+     */
     @Override
-    public TrackerBundleParams convert( TrackerBundleParams bundle )
+    public TrackerBundleParams convert( TrackerBundleParams dataBundle )
     {
-        // pre assign UIDs to entities, if UIDs are missing
-        generateUid( bundle );
+        Map<String, TrackedEntity> trackedEntityMap = new HashMap<>();
+        Map<String, Enrollment> enrollmentHashMap = new HashMap<>();
+        Map<String, Event> eventHashMap = new HashMap<>();
+        Map<String, Relationship> relationshipHashMap = new HashMap<>();
 
-        if ( hasNestedStructure( bundle ) )
+        // Extract all enrollments and relationships, and set parent reference.
+        for ( TrackedEntity te : dataBundle.getTrackedEntities() )
         {
-            flattenPayload( bundle );
-        }
+            updateTrackedEntityReferences( te );
+            trackedEntityMap.put( te.getTrackedEntity(), te );
 
-        return bundle;
-    }
+            extractEnrollments( te )
+                .forEach( enrollment -> enrollmentHashMap.put( enrollment.getEnrollment(), enrollment ) );
 
-    private void flattenPayload( TrackerBundleParams bundle )
-    {
+            extractRelationships( te )
+                .forEach( relationship -> relationshipHashMap.put( relationship.getRelationship(), relationship ) );
+        }
 
-        List<Event> events = new ArrayList<>();
+        // Set UID for all enrollments and notes
+        dataBundle.getEnrollments().stream()
+            .peek( enrollment -> updateEnrollmentReferences( enrollment, enrollment.getTrackedEntity() ) )
```

Comment:
As we discussed, make sure that we only generate UID for `enrollment.getEnrollment()`
The `trackedEntity` property should be left empty if it's not present in the original payload

```
Enrollment:
updateReference(enrollment.getEnrollment)
enrollment.getTei -> isEmpty() ? null : tei
```